### ISSUE TITLE
generateKubeconfig expiring after 16hours

### ISSUE THREAD

**Issue by mattmattox** [rancher/rancher#26121]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
- Login into the API
https://${RANCHER}/v3-public/activeDirectoryProviders/activedirectory?action=login
- Generate kubeconfig
https://${RANCHER}/v3/clusters/${CLUSTERID}?action=generateKubeconfig > output1
- Wait a few seconds
- Generate another kubeconfig
https://${RANCHER}/v3/clusters/${CLUSTERID}?action=generateKubeconfig > output2
- diff output1 output2 (The tokens will match)

**Result:**
Once the kubeconfig is generated for a user. They will keep getting the same token until it expires in 16 hours at which point they will get a new token. This breaks scheduled jobs due to the token expiring even tho they are requesting the kubeconfig at the bring of each script.

Example:
Scheduled job that runs 15/mins. It will get a valid token on the first job, then get the same token for the next 16 hours. But if the token expires while a job is running (16Hrs), then that job will fail until a new generateKubeconfig is created IE on the next run. This will cause jobs/builds/monitoring to fail every 16 hours.

**Other details that may be helpful:**
Possible workarounds / fixes
A) Add a flag to generateKubeconfig to create a non-expiring token for long running processing (CI/CD, monitoring, etc)
B) Add a flag to set the ttl for that token
C) Create a new token every time generateKubeconfig is called.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.3.5
- Installation option (single install/HA): HA

**Comment by kinarashah** [comment]
@mattmattox We recently introduced a feature (2.4.7) that allows admins to enforce TTLs on Kubeconfig tokens. Could you confirm if that works for this use case? 

Flow would be something like: 
- admins go and update 2 settings, kubeconfig-generate-token=false (so rancher doesn't generate token) and kubeconfig-token-ttl-minutes=1440 for instance 
- Kubeconfig no longer has this token stored in the file. Running kubectl now requires rancher CLI to be present in your path and kubectl will ask user to login to Rancher to get a new token. 

Here's the doc link for more details, https://github.com/rancher/docs/pull/2647

**Comment by ansilh** [comment]
@kinarashah 
Here are the steps I followed 

Set `kubeconfig-generate-token=false` & `kubeconfig-token-ttl-minutes=1` ([doc](https://rancher.com/docs/rancher/v2.x/en/api/api-tokens/#setting-ttl-on-kubeconfig-tokens))

- Login into the API
`https://${RANCHER}/v3-public/activeDirectoryProviders/activedirectory?action=login`

- Generate kubeconfig
`https://${RANCHER}/v3/clusters/${CLUSTERID}?action=generateKubeconfig | python2 -c "import sys, json; print json.load(sys.stdin)['config']" >cluster.yaml`

- Download and place rancher CLI where kubectl is getting executed 
- Execute `kubectl get nodes --kubeconfig cluster.yaml`
- User will be asked to login as kubectl calls rancher cli when we execute the command first time
- After successful authentication, a token will be created and stored in `.cache/token/cluster_c-id.json` with ttl of 1 minute.
- Successive kubectl commands will succeed till token expiration.
- After one minute, the token will expire and kubectl command (indirectly rancher cli) will again prompt for authentication.

All looks good here. 

Now let's take the below scenario ;

 - We executed `kubectl get nodes` and that will call rancher cli to get token. 
 - The existing token stored in `.cache/token/cluster_c-id.json` will be validated by rancher cli and pass that to kubectl
 - Suppose there is a delay in passing the token to kubectl or the token expired before it reaches kubectl.

Now the kubectl call will fail and we run into the same problem which @mattmattox explained earlier. 
Moreover, we are dealing with a global setting and if some users expecting token by default in kubeconfig and expecting a different ttl (like 5 minutes) then that scenario will also fail based on the steps we ([documented](https://rancher.com/docs/rancher/v2.x/en/api/api-tokens/#setting-ttl-on-kubeconfig-tokens)).

Please correct me if my understanding of the auth/token workflow is not right.

**Comment by kinarashah** [comment]
@ansilh It's true, a delay in passing the token to kubectl or if it expires after CLI's validation and before auth check, the process will fail - and retry will force the user to login again. 

If this is an automated script which requires 4-5 mins TTL, a workaround that I can think of is: 
- the script gets the kubeconfig token from API and if it's close to expiry, deletes the token and calls generateKubeconfig again. 

Is this workaround feasible? 

Else, based on your scenario, we'll need to implement the following enhancements, what do you think? 
- custom TTL for kubeconfig tokens per cluster 
- CLI's validation check can add some time buffer when checking if the token's expired (this will resolve the use case you mention) 
- introduce env vars for username, password (would work for active directory, not SAML) in CLI, so when kubectl fails for some reason, the next call can use env credentials and get the new token without requiring user intervention (this might help resolve the automation case you talk about?)

**Comment by ansilh** [comment]
Yes, the logic is part of the automation script which needs a token that lives till the job completes (job may have multiple kubectl calls) 
The workaround is feasible, but we can enhance the experience for the end-user.

Here are my thoughts on the enhancement.

> * custom TTL for kubeconfig tokens per cluster
* Allow users to generate kubeconfig with custom TTL instead of a cluster-wide TTL

> * CLI's validation check can add some time buffer when checking if the token's expired (this will resolve the use case you mention)

* CLI can add a validation check for the time left for the expiry of the token (maybe a flag with tunable buffer time - if that is what meant by the buffer time)

> * introduce env vars for username, password (would work for active directory, not SAML) in CLI, so when kubectl fails for some reason, the next call can use env credentials and get the new token without requiring user intervention (this might help resolve the automation case you talk about?)

* We cannot afford kubectl call failure due to expired token. 

Ideally, we should let the user decide the TTL of the token when the user request kubeconfig.
Also, add a feature to generate a new token whenever the user request for kubeconfig with no TTL in the request and let the global flag to decide the default TTL. This may also need a max value for the number of tokens each user can generate to avoid token sprawl. 

Thanks for your inputs.
